### ISSUE TITLE
Attacks not working with PyTorch model

### ISSUE THREAD

**Issue by HadiAbdullah** [bethgelab/foolbox#153]
Most of the attacks are not working with the PyTorch model. The attack methods return a none type, when in fact they should return adversarial images. The only attack that has worked is the LBFGSAttack.

# Setup model
`model_conv = torchvision.models.resnet50(pretrained=True)`
`model_conv = model_conv.eval()`
`model = foolbox.models.PyTorchModel(model_conv, bounds=(0, 255),num_classes=1000, cuda=False)`

# Get image to perturb
`image, label = foolbox.utils.imagenet_example()`
`image = np.asarray(image.reshape([1,3,224,224]))[0]`
`label = np.argmax(model.predictions(image))`
`
# Attack criteria 
`criterion = foolbox.criteria.Misclassification()`
`attack = at.IterativeGradientAttack(model,criterion=criterion)`
`adversarial = attack(image, label)`

**Comment by jonasrauber** [comment]
This code contains quite a number of mistakes. Fixing them will solve the problem ;-)

- to change an image from 224x224x3 to 3x224x224, you cannot use reshape, but you need to transpose axes!
- the torchvision models expects inputs that are preprocessed (mean and std normalization, see torchvision documentation)
- the bounds should be (0, 1)
- the image values need to be mapped to [0, 1] accordingly
- you should use the true label, not the one predicted by the, then you might have caught these mistakes earlier

Please have a look at the PyTorch example in the documentation: [https://foolbox.readthedocs.io/en/latest/user/examples.html](https://foolbox.readthedocs.io/en/latest/user/examples.html#creating-an-untargeted-adversarial-for-a-pytorch-model)

**Comment by vipinpillai** [comment]
@jonasrauber It is actually recommended to use the predicted label instead of the true label due to the label leaking effect mentioned in [Adversarial Machine Learning at Scale](https://arxiv.org/pdf/1611.01236.pdf)

**Comment by jonasrauber** [comment]
@vipinpillai That does not apply here. If you specify the true label, Foolbox will only actually run the attack, if the original image is classified correctly, i.e. if the predicted label is the same as the original label anyway. If they are different, Foolbox won't run the attack. The consequence of specifying the predicted label instead of the true one to Foolbox is that you might get an "adversarial" that's classified correctly — certainly not what one wants.
The only conclusion you should take from the paper you mention is that you shouldn't take (rely on) single-step adversarial attacks because high robustness against them doesn't tell you much about the model's actual robustness (these attacks are just too weak).

**Comment by vipinpillai** [comment]
@jonasrauber Yes I agree that single-step adversarial attacks are indeed weak. But the label-leaking effect shows that the network can start exploiting the ground-truth specific information used to construct the adversarial noise. There hasn't been any work which shows that label-leaking is not possible for iterative attacks.
